File: csharp-project/RekindleBookStoreAqa/RekindleTaf/Controllers/BookstoreController.cs
using System.Collections.Generic;
using System.Net;
using System.Text.Json.Nodes;
using System.Threading.Tasks;
using FluentAssertions;
using RekindleTaf.Configuration;
using RekindleTaf.Configuration.Rest;
using RekindleTaf.Model;
using RestSharp;

namespace RekindleTaf.Controllers;

public class BookstoreController(IEnvConfig envConfig, IRestConfig restConfig) : IBookstoreController
{
    public async Task<RestResponse> Get(string param = "")
    {
        var request = new RestRequest(envConfig.BookstoreEndpoints().BookstoresById)
            .AddUrlSegment("bookstoreId", param);
        return await restConfig.RekindleClient().ExecuteGetAsync(request);
    }

    public Bookstore GetSuccessfully(string param = "")
    {
        var response = Get(param).Result;
        response.StatusCode.Should().Be(HttpStatusCode.OK);
        return restConfig.Deserialize<Bookstore>(response);
    }

    public async Task<RestResponse> GetAll(string param = "")
    {
        var request = new RestRequest(envConfig.BookstoreEndpoints().Bookstores);
        return await restConfig.RekindleClient().ExecuteGetAsync(request);
    }

    public List<Bookstore> GetAllSuccessfully(string param = "")
    {
        var response = GetAll(param).Result;
        response.StatusCode.Should().Be(HttpStatusCode.OK);
        return restConfig.Deserialize<List<Bookstore>>(response);
    }

    public async Task<RestResponse> Post(Bookstore param)
    {
        var request = new RestRequest(envConfig.BookstoreEndpoints().Bookstores, Method.Post)
            .AddJsonBody(param);
        return await restConfig.RekindleClient().ExecutePostAsync(request);
    }

    public Bookstore PostSuccessfully(Bookstore param)
    {
        var response = Post(param).Result;
        response.StatusCode.Should().Be(HttpStatusCode.Created);
        var responseData = restConfig.Deserialize<JsonNode>(response);
        param.Id = responseData.ToString();
        return param;
    }

    public async Task<RestResponse> Put(Bookstore param)
    {
        var request = new RestRequest(envConfig.BookstoreEndpoints().BookstoresById, Method.Put)
            .AddUrlSegment("bookstoreId", param.Id)
            .AddJsonBody(param);
        return await restConfig.RekindleClient().ExecutePutAsync(request);
    }

    public void PutSuccessfully(Bookstore param)
    {
        var response = Put(param).Result;
        response.StatusCode.Should().Be(HttpStatusCode.NoContent);
    }

    public async Task<RestResponse> Delete(Bookstore param)
    {
        var request = new RestRequest(envConfig.BookstoreEndpoints().BookstoresById, Method.Delete)
            .AddUrlSegment("bookstoreId", param.Id);
        return await restConfig.RekindleClient().ExecuteDeleteAsync(request);
    }

    public void DeleteSuccessfully(Bookstore param)
    {
        var response = Delete(param).Result;
        response.StatusCode.Should().Be(HttpStatusCode.NoContent);
    }
}

File: typescript-project/rekindle-book-store-aqa/src/rest/controllers/bookstore-controller.ts
import {Bookstore} from "@/model/domain-model";
import {AxiosInstance, AxiosResponse, HttpStatusCode} from "axios";
import {BookstoreEndpoints} from "@/rest/Endpoint";
import {BookstoreCrudController} from "@/rest/controllers/rest-controllers";
import {expect} from "@jest/globals";
import {RekindleClient} from "@/rest/rekindle-rest-client";

export class BookstoreController implements BookstoreCrudController {

    private readonly client: AxiosInstance;

    constructor(client: RekindleClient) {
        this.client = client.getAxios();
    }

    async get(param?: string): Promise<AxiosResponse> {
        return await this.client.request({
            method: 'GET',
            url: BookstoreEndpoints.bookstoresById({bookstoreId: param})
        });
    }

    async getSuccessfully(param?: string): Promise<Bookstore> {
        try {
            const resp = await this.get(param);
            expect(resp.status).toBe(HttpStatusCode.Ok);
            return resp.data;
        } catch (err) {
            throw err;
        }
    }

    async getAll(param?: string): Promise<AxiosResponse> {
        return await this.client.request({
            method: 'GET',
            url: BookstoreEndpoints.bookstores
        });
    }

    async getAllSuccessfully(param?: string): Promise<Bookstore[]> {
        try {
            const resp = await this.getAll(param);
            expect(resp.status).toBe(HttpStatusCode.Ok);
            return resp.data
        } catch (err) {
            throw err;
        }
    }

    async post(param: Bookstore): Promise<AxiosResponse> {
        return await this.client.request({
            method: 'POST',
            url: BookstoreEndpoints.bookstores,
            data: param
        });
    }

    async postSuccessfully(param: Bookstore): Promise<Bookstore> {
        try {
            let bookstore = param;
            const resp = await this.post(param);
            expect(resp.status).toBe(HttpStatusCode.Created);
            bookstore.id = resp.data
            return bookstore;
        } catch (err) {
            throw err;
        }
    }

    async put(param: Bookstore): Promise<AxiosResponse> {
        return await this.client.request({
            method: 'PUT',
            url: BookstoreEndpoints.bookstoresById({bookstoreId: param.id}),
            data: param
        });
    }

    async putSuccessfully(param: Bookstore): Promise<void> {
        try {
            const resp = await this.put(param);
            expect(resp.status).toBe(HttpStatusCode.NoContent);
        } catch (err) {
            throw err;
        }
    }

    async delete(param: Bookstore): Promise<AxiosResponse> {
        return await this.client.request({
            method: 'DELETE',
            url: BookstoreEndpoints.bookstoresById({bookstoreId: param.id})
        });
    }

    async deleteSuccessfully(param: Bookstore): Promise<void> {
        try {
            const resp = await this.delete(param);
            expect(resp.status).toBe(HttpStatusCode.NoContent);
        } catch (err) {
            throw err;
        }
    }
}